extensions/stackdriver/edges/edge_reporter.cc

```diff
@@ -0,0 +1,141 @@
+
+#ifndef NULL_PLUGIN
+#include "api/wasm/cpp/proxy_wasm_intrinsics.h"
+#else
+#include "extensions/common/wasm/null/null_plugin.h"
+#endif
+
+namespace Extensions {
+namespace Stackdriver {
+namespace Edges {
+
+using google::cloud::meshtelemetry::v1alpha1::ReportTrafficAssertionsRequest;
+using google::cloud::meshtelemetry::v1alpha1::TrafficAssertion;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_GRPC;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTP;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_HTTPS;
+using google::cloud::meshtelemetry::v1alpha1::
+    TrafficAssertion_Protocol_PROTOCOL_TCP;
+using google::cloud::meshtelemetry::v1alpha1::WorkloadInstance;
+using google::protobuf::util::TimeUtil;
+
+std::unique_ptr<WorkloadInstance> instanceFromMetadata(
+    const ::wasm::common::NodeInfo& node_info) {
+  auto instance = std::make_unique<WorkloadInstance>();
+  // TODO(douglas-reid): support more than just kubernetes instances
+  instance->set_uid("kubernetes://" + node_info.name() + "." +
+                    node_info.namespace_());
+  // TODO(douglas-reid): support more than just GCP ?
+  instance->set_location(
+      node_info.platform_metadata().at(Common::kGCPClusterLocationKey));
+  instance->set_cluster_name(
+      node_info.platform_metadata().at(Common::kGCPClusterNameKey));
+  instance->set_owner_uid(node_info.owner());
+  instance->set_workload_name(node_info.workload_name());
+  instance->set_workload_namespace(node_info.namespace_());
+  return instance;
+};
+
+EdgeReporter::EdgeReporter(
+    const ::wasm::common::NodeInfo& local_node_info,
+    std::unique_ptr<MeshEdgesServiceClient> edges_client) {
+  current_request_ = std::make_unique<ReportTrafficAssertionsRequest>();
+
+  const auto& project_id =
+      local_node_info.platform_metadata().at(Common::kGCPProjectKey);
+  current_request_->set_parent("projects/" + project_id);
+
+  // TODO(dougreid): figure out how to get the real mesh uid here.
+  // Using: //.../projects/<project_id>/<location>/meshes/<cluster> as a
+  // placeholder.
+  std::string location =
+      local_node_info.platform_metadata().at(Common::kGCPClusterLocationKey);
+  std::string cluster =
+      local_node_info.platform_metadata().at(Common::kGCPClusterNameKey);
+  std::string mesh_uid = "unknown";
+  current_request_->set_mesh_uid(
+      "//cloudresourcemanager.googleapis.com/projects/" + project_id + "/" +
+      location + "/meshes/" + cluster);
+
+  node_instance_ = instanceFromMetadata(local_node_info);
+
+  edges_client_ = std::move(edges_client);
+};
+
+// ONLY inbound
+void EdgeReporter::addEdge(const ::Wasm::Common::RequestInfo& request_info,
+                           const ::wasm::common::NodeInfo& peer_node_info) {
+  auto* traffic_assertions = current_request_->mutable_traffic_assertions();
+  auto* edge = traffic_assertions->Add();
```

Comment:
This using too many allocs and adding too many samples.
Look at https://github.com/istio/proxy/blob/master/extensions/stats/plugin.cc#L118
you can use a map like in the stats plugin, define a key that includes important, or *all* bits you need to use.

Reply:
I was avoiding optimizations to start, but have since added a TrafficAssertionsRequest-bound cache that will prevent adding the same edge over and over. there is no cache across reports to the telemetry service yet, but that can be added later.

let me know what you think.